cache.go

```diff
@@ -0,0 +1,192 @@
+package libovsdb
+
+import (
+	"reflect"
+	"sync"
+)
+
+// RowCache is a collections of Rows hashed by UUID
+type RowCache struct {
+	cache map[string]Row
+	mutex sync.Mutex
+}
+
+// Row returns one row the from the cache by uuid
+func (r *RowCache) Row(uuid string) *Row {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if row, ok := r.cache[uuid]; ok {
+		return &row
+	}
+	return nil
+}
+
+// Rows returns a list of row UUIDs as strings
+func (r *RowCache) Rows() []string {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	var result []string
+	for k := range r.cache {
+		result = append(result, k)
+	}
+	return result
+}
+
+func newRowCache() *RowCache {
+	return &RowCache{
+		cache: make(map[string]Row),
+		mutex: sync.Mutex{},
+	}
+}
+
+// EventHandler can handle events that happen to cache rowects
+type EventHandler interface {
+	OnAdd(table string, row Row)
+	OnUpdate(table string, old Row, new Row)
+	OnDelete(table string, row Row)
+}
+
+// EventHandlerFuncs is a wrapper for the EventHandler interface
+// It allows a caller to only implement the functions they need
+type EventHandlerFuncs struct {
+	AddFunc    func(table string, row Row)
+	UpdateFunc func(table string, old Row, new Row)
+	DeleteFunc func(table string, row Row)
+}
+
+// OnAdd calls AddFunc if it is not nil
+func (e *EventHandlerFuncs) OnAdd(table string, row Row) {
+	if e.AddFunc != nil {
+		e.AddFunc(table, row)
+	}
+}
+
+// OnUpdate calls UpdateFunc if it is not nil
+func (e *EventHandlerFuncs) OnUpdate(table string, old, new Row) {
+	if e.UpdateFunc != nil {
+		e.UpdateFunc(table, old, new)
+	}
+}
+
+// OnDelete calls DeleteFunc if it is not nil
+func (e *EventHandlerFuncs) OnDelete(table string, row Row) {
+	if e.DeleteFunc != nil {
+		e.DeleteFunc(table, row)
+	}
+}
+
+// TableCache is a collection of TableCache hashed by database name
```

Comment:
Shouldn't it be: "TableCache is a collection of RowCache hashed by the table name"?

Reply:
Ack. I'll fix that